langwatch/src/features/onboarding/components/sections/model-provider/ModelProviderSetup.tsx

```diff
@@ -0,0 +1,312 @@
+      refetch().catch(err => easyCatch(err, "ModelProviderSetup.onSuccess"));
+    },
+  });
+
+  const { fields: derivedFields } = useModelProviderFields(backendModelProviderKey);
+  const [openAiValidationError, setOpenAiValidationError] = useState<string>();
+  const [fieldErrors, setFieldErrors] = useState<Record<string, string>>({});
+
+  useEffect(() => {
+    setOpenAiValidationError(void 0);
+    setFieldErrors({});
+  }, [modelProviderKey]);
+
+  const isOpenAiProvider = backendModelProviderKey === "openai";
+
+  const handleCustomKeyChange = useCallback(
+    (key: string, value: string) => {
+      actions.setCustomKey(key, value);
+    },
+    [actions],
+  );
+
+  const handleFieldErrorClear = useCallback((key: string) => {
+    setFieldErrors((previous) => {
+      if (!previous[key]) {
+        return previous;
+      }
+      const updated = { ...previous };
+      delete updated[key];
+      return updated;
+    });
+  }, []);
+
+  const handleOpenAiValidationClear = useCallback(() => {
+    setOpenAiValidationError(void 0);
+  }, []);
+
+  const validateOpenAi = useCallback(() => {
+    if (!isOpenAiProvider) return true;
+
+    const apiKey = state.customKeys.OPENAI_API_KEY?.trim() ?? "";
+    const baseUrl = state.customKeys.OPENAI_BASE_URL?.trim() ?? "";
+
+    // Both empty
+    if (!apiKey && !baseUrl) {
+      setOpenAiValidationError("Either API Key or Base URL must be provided");
+      return false;
+    }
+
+    // Base URL is set to default OpenAI URL, but no API key
+    if (baseUrl === OPENAI_DEFAULT_BASE_URL && !apiKey) {
+      setOpenAiValidationError(
+        "API Key is required when using the default OpenAI base URL",
+      );
+      return false;
+    }
+
+    handleOpenAiValidationClear();
+    return true;
+  }, [isOpenAiProvider, state.customKeys, handleOpenAiValidationClear]);
+
+  const handleSaveAndContinue = useCallback(() => {
+    if (!validateOpenAi()) {
+      return;
+    }
+
+    // Clear previous errors
+    setFieldErrors({});
+    handleOpenAiValidationClear();
+
+    // Validate keys according to schema before submitting
+    const providerDefinition = backendModelProviderKey
+      ? modelProvidersRegistry[backendModelProviderKey]
+      : void 0;
+
+    if (providerDefinition?.keysSchema) {
+      const keysSchema = z.union([
+        providerDefinition.keysSchema,
+        z.object({ MANAGED: z.string() }),
+      ]);
+
+      const keysToValidate: Record<string, unknown> = { ...state.customKeys };
+      const result = keysSchema.safeParse(keysToValidate);
+
+      if (!result.success) {
+        // Parse the Zod error to get field-specific errors
+        const parsedErrors = parseZodFieldErrors(
+          result.error as ZodErrorStructure,
+        );
+        setFieldErrors(parsedErrors);
+        return;
+      }
+    }
+
+    void actions
+      .setEnabled(true)
+      .then(() => actions.submit())
+      .catch(err => easyCatch(err, "ModelProviderSetup.handleSaveAndContinue"));
```

Comment:
Nice! You did it! But shouldn't the error have the stack already? why pass the string? Seems like it adds unnecessary complexity to use -- like if you're going to write it out, why not just use the logger itself:

```
err => {
   logger.error(err, '....')
```

is basically the same amount of code

Reply:
yeah good call, will do!